File: aetherctl/src/openspec/security.ts
import { resolve, isAbsolute } from "node:path";

// Security constants
export const ALLOWED_COMMANDS = [
  "npm",
  "pnpm",
  "yarn",
  "git",
  "make",
  "python",
  "python3",
  "pytest",
  "node",
  "tsc",
  "eslint",
  "cargo",
  "go",
  "mvn",
  "gradle",
];

export const MAX_TASKS_FILE_SIZE = 1024 * 1024; // 1MB
export const MAX_LINE_LENGTH = 10000;
export const MAX_TASK_COUNT = 1000;
export const MAX_CMD_BUFFER_SIZE = 50 * 1024 * 1024; // 50MB
export const STALE_LOCK_AGE_MS = 300000; // 5 minutes

/**
 * Sanitize path component to prevent path traversal attacks.
 * @throws Error if path component is invalid
 */
export function sanitizePathComponent(component: string): string {
  if (!component || component.trim().length === 0) {
    throw new Error("Path component cannot be empty");
  }

  // Reject path traversal attempts
  if (
    component.includes("..") ||
    component.includes("/") ||
    component.includes("\\") ||
    component.includes("\0")
  ) {
    throw new Error(`Invalid path component: ${component}`);
  }

  return component.trim();
}

/**
 * Safely construct a path, ensuring it stays within the base directory.
 * @throws Error if path traversal is detected
 */
export function safePath(base: string, ...components: string[]): string {
  if (!isAbsolute(base)) {
    throw new Error(`Base path must be absolute: ${base}`);
  }

  // Sanitize all components
  const sanitized = components.map(sanitizePathComponent);

  // Construct and verify final path
  const fullPath = resolve(base, ...sanitized);
  const normalizedBase = resolve(base);

  if (!fullPath.startsWith(normalizedBase + "/") && fullPath !== normalizedBase) {
    throw new Error(`Path traversal detected: ${fullPath}`);
  }

  return fullPath;
}

/**
 * Validate command against whitelist and extract command + args safely.
 * @throws Error if command is not allowed
 */
export function validateCommand(cmd: string): { command: string; args: string[] } {
  if (!cmd || cmd.trim().length === 0) {
    throw new Error("Command cannot be empty");
  }

  const trimmed = cmd.trim();

  // Check for shell injection patterns
  const dangerousPatterns = [
    /[;&|`$()]/g, // Shell metacharacters
    /\n|\r/g, // Newlines
    />\s*\/dev/g, // Device access
    /rm\s+-rf\s+\//g, // Destructive commands
  ];

  for (const pattern of dangerousPatterns) {
    if (pattern.test(trimmed)) {
      throw new Error(`Command contains dangerous pattern: ${trimmed}`);
    }
  }

  // Parse command and args
  const parts = trimmed.split(/\s+/).filter((p) => p.length > 0);
  if (parts.length === 0) {
    throw new Error("Command cannot be empty after parsing");
  }

  const command = parts[0];
  const args = parts.slice(1);

  // Check whitelist
  if (!ALLOWED_COMMANDS.includes(command)) {
    throw new Error(
      `Command not in whitelist: ${command}. Allowed: ${ALLOWED_COMMANDS.join(", ")}`
    );
  }

  return { command, args };
}

/**
 * Sanitize log message to prevent log injection.
 */
export function sanitizeLogMessage(s: string): string {
  if (!s) return "";
  return s.replace(/[\r\n]/g, "\\n").replace(/\t/g, "\\t").slice(0, 10000);
}

/**
 * Redact sensitive information from error messages.
 */
export function redactSensitiveInfo(s: string): string {
  if (!s) return "";

  let result = s;

  // Redact common sensitive patterns
  result = result.replace(/password[=:]\s*\S+/gi, "password=***");
  result = result.replace(/token[=:]\s*\S+/gi, "token=***");
  result = result.replace(/api[_-]?key[=:]\s*\S+/gi, "api_key=***");
  result = result.replace(/secret[=:]\s*\S+/gi, "secret=***");
  result = result.replace(/\/home\/[^/\s]+/g, "/home/***");
  result = result.replace(/\/Users\/[^/\s]+/g, "/Users/***");

  return result;
}

/**
 * Validate task markdown content.
 * @throws Error if content is invalid
 */
export function validateTasksContent(content: string): void {
  if (content.length > MAX_TASKS_FILE_SIZE) {
    throw new Error(`Tasks file too large: ${content.length} bytes (max: ${MAX_TASKS_FILE_SIZE})`);
  }

  const lines = content.split("\n");
  if (lines.length > MAX_TASK_COUNT) {
    throw new Error(`Too many lines: ${lines.length} (max: ${MAX_TASK_COUNT})`);
  }

  for (let i = 0; i < lines.length; i++) {
    if (lines[i].length > MAX_LINE_LENGTH) {
      throw new Error(`Line ${i + 1} too long: ${lines[i].length} chars (max: ${MAX_LINE_LENGTH})`);
    }
  }
}

/**
 * Sanitize environment variables for subprocess execution.
 */
export function sanitizeEnv(env: NodeJS.ProcessEnv): NodeJS.ProcessEnv {
  const safe: NodeJS.ProcessEnv = {
    PATH: env.PATH,
    HOME: env.HOME,
    USER: env.USER,
    LANG: env.LANG,
    LC_ALL: env.LC_ALL,
    TERM: env.TERM,
  };

  // Remove undefined values
  Object.keys(safe).forEach((key) => {
    if (safe[key] === undefined) {
      delete safe[key];
    }
  });

  return safe;
}

File: nanobot/agent/taskmode/security.py
"""Command security for Task-Mode execution.

Ports the safety checks from aetherctl/src/openspec/security.ts and
reuses patterns from nanobot/agent/tools/shell.py.
"""

from __future__ import annotations

import os
import re
import shlex
from dataclasses import dataclass

# ---------------------------------------------------------------------------
# Whitelist of allowed base commands
# ---------------------------------------------------------------------------

ALLOWED_COMMANDS: list[str] = [
    # Package managers
    "npm", "pnpm", "yarn", "pip", "pip3", "uv", "poetry", "pipx",
    # Version control
    "git",
    # Build
    "make", "cmake", "cargo", "go", "mvn", "gradle",
    # Runtimes / interpreters
    "python", "python3", "node", "deno", "bun",
    # Linters / formatters
    "tsc", "eslint", "prettier", "ruff", "black", "isort", "mypy", "pylint",
    # Test runners
    "pytest", "vitest", "jest",
    # AI CLI tools
    "claude", "codex", "gemini", "opencode",
    # Common utilities
    "echo", "cat", "ls", "cp", "mv", "mkdir", "touch", "chmod",
    "grep", "find", "sed", "awk", "head", "tail", "wc", "sort", "uniq",
    "curl", "wget",
    "docker", "docker-compose",
    "bash", "sh", "zsh",
]

# ---------------------------------------------------------------------------
# Dangerous patterns — block regardless of whitelist
# ---------------------------------------------------------------------------

_DANGEROUS_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\brm\s+-[rf]{1,2}\s+/(?!\S*/openspec)"),   # rm -rf / (but allow within openspec)
    re.compile(r"\bdel\s+/[fq]", re.IGNORECASE),             # Windows del /f
    re.compile(r"\b(?:format|mkfs|diskpart)\b", re.IGNORECASE),
    re.compile(r"\bdd\s+if="),                                 # dd
    re.compile(r">\s*/dev/sd"),                                # write to raw disk
    re.compile(r"\b(?:shutdown|reboot|poweroff|init\s+0)\b"),  # power control
    re.compile(r":\(\)\s*\{.*\};\s*:"),                        # fork bomb
    re.compile(r"\bchmod\s+777\s+/"),                          # recursive world-writable root
    re.compile(r"\bcurl\b.*\|\s*(?:ba)?sh"),                   # curl | sh
    re.compile(r"\bwget\b.*\|\s*(?:ba)?sh"),                   # wget | sh
]

# Shell meta-characters that indicate injection risk
_INJECTION_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"`"),            # backtick substitution
    re.compile(r"\$\("),         # $() subshell
    re.compile(r"\n|\r"),        # newlines
    re.compile(r">\s*/dev/"),    # device writes
]


@dataclass
class CommandValidation:
    """Result of command validation."""
    ok: bool
    command: str = ""
    args: list[str] | None = None
    error: str | None = None


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def validate_command(cmd: str) -> CommandValidation:
    """Validate a command string from tasks.md.

    Checks:
    1. Not empty
    2. No dangerous patterns (rm -rf /, fork bomb, etc.)
    3. No shell injection patterns (backticks, $(), newlines)
    4. Base command is in the whitelist
    5. Parses into command + args for shell=False execution

    Returns a ``CommandValidation`` with either parsed command/args or
    an error message.
    """
    if not cmd or not cmd.strip():
        return CommandValidation(ok=False, error="Command cannot be empty")

    trimmed = cmd.strip()

    # Check dangerous patterns
    for pattern in _DANGEROUS_PATTERNS:
        if pattern.search(trimmed):
            return CommandValidation(
                ok=False,
                error=f"Command blocked (dangerous pattern): {trimmed[:100]}",
            )

    # Check injection patterns
    for pattern in _INJECTION_PATTERNS:
        if pattern.search(trimmed):
            return CommandValidation(
                ok=False,
                error=f"Command blocked (shell injection pattern): {trimmed[:100]}",
            )

    # Reject chaining operators: ; && ||
    # But allow them inside quoted strings by using shlex
    # Simple heuristic: check unquoted occurrences
    if _has_unquoted_chaining(trimmed):
        return CommandValidation(
            ok=False,
            error=f"Command blocked (chaining operators not allowed): {trimmed[:100]}",
        )

    # Parse command into parts
    try:
        parts = shlex.split(trimmed)
    except ValueError as e:
        return CommandValidation(
            ok=False,
            error=f"Command parse error: {e}",
        )

    if not parts:
        return CommandValidation(ok=False, error="Command is empty after parsing")

    base_cmd = parts[0]

    # Allow path-qualified commands if the basename is whitelisted
    # e.g. /usr/bin/python3 -> python3, ./node_modules/.bin/eslint -> eslint
    base_name = os.path.basename(base_cmd)

    if base_name not in ALLOWED_COMMANDS:
        return CommandValidation(
            ok=False,
            error=f"Command not in whitelist: `{base_name}`. "
                  f"Allowed: {', '.join(sorted(ALLOWED_COMMANDS)[:20])}...",
        )

    return CommandValidation(
        ok=True,
        command=base_cmd,
        args=parts[1:],
    )


def sanitize_env() -> dict[str, str]:
    """Build a minimal safe environment for subprocess execution."""
    keys = ["PATH", "HOME", "USER", "LANG", "LC_ALL", "TERM", "SHELL",
            "TMPDIR", "XDG_RUNTIME_DIR"]
    return {k: os.environ[k] for k in keys if k in os.environ}


def sanitize_log_message(s: str, max_len: int = 10_000) -> str:
    """Strip newlines/tabs to prevent log injection."""
    if not s:
        return ""
    return s.replace("\r", "\\r").replace("\n", "\\n").replace("\t", "\\t")[:max_len]


def redact_sensitive(s: str) -> str:
    """Mask passwords, tokens, keys, and user paths in log output."""
    if not s:
        return ""
    result = s
    result = re.sub(r"(?i)password[=:]\s*\S+", "password=***", result)
    result = re.sub(r"(?i)token[=:]\s*\S+", "token=***", result)
    result = re.sub(r"(?i)api[_-]?key[=:]\s*\S+", "api_key=***", result)
    result = re.sub(r"(?i)secret[=:]\s*\S+", "secret=***", result)
    result = re.sub(r"/Users/[^/\s]+", "/Users/***", result)
    result = re.sub(r"/home/[^/\s]+", "/home/***", result)
    return result


# ---------------------------------------------------------------------------
# Internal
# ---------------------------------------------------------------------------

def _has_unquoted_chaining(cmd: str) -> bool:
    """Check for ;, &&, || outside of quotes."""
    in_single = False
    in_double = False
    i = 0
    n = len(cmd)
    while i < n:
        c = cmd[i]
        if c == "'" and not in_double:
            in_single = not in_single
        elif c == '"' and not in_single:
            in_double = not in_double
        elif not in_single and not in_double:
            if c == ";":
                return True
            if c == "&" and i + 1 < n and cmd[i + 1] == "&":
                return True
            if c == "|" and i + 1 < n and cmd[i + 1] == "|":
                return True
        i += 1
    return False
